recommendation/widedeep/widedeep.py

```diff
@@ -0,0 +1,119 @@
+
+        full_features_names = numeric_feature_names + string_feature_names + ["label"]
+        feature_datatypes = [tf.io.FixedLenSequenceFeature([], tf.float32, default_value=0.0, allow_missing=True)] + \
+                            [tf.io.FixedLenSequenceFeature([], tf.int64, default_value=0, allow_missing=True)] + \
+                            [tf.io.FixedLenSequenceFeature([], tf.int64, default_value=0, allow_missing=True)]
+
+        def _parse_function(proto):
+            f = collections.OrderedDict(
+                zip(full_features_names, feature_datatypes))
+            parsed_features = tf.io.parse_example(proto, f)
+            parsed_feature_vals_num = [tf.reshape(
+                parsed_features["numeric_1"], shape=[-1, 13])]
+            parsed_feature_vals_str = [tf.reshape(
+                parsed_features["string_1"], shape=[-1, 2]) for i in string_feature_names]
+            parsed_feature_vals = parsed_feature_vals_num + parsed_feature_vals_str
+
+            # labels
+            parsed_feature_vals_label = [tf.reshape(parsed_features[i], shape=[-1]) for i in ["label"]]
+            parsed_feature_vals = parsed_feature_vals + parsed_feature_vals_label
+
+            return parsed_feature_vals
+
+        # Extract lines from input files using the Dataset API.
+        dataset = tf.data.TFRecordDataset([self.tfrecords_path])
+        if shuffle:
+            dataset = dataset.shuffle(buffer_size=20000)
+        dataset = dataset.batch(self.batch_size)
+        dataset = dataset.map(_parse_function, num_parallel_calls=28)
+        dataset = dataset.prefetch(self.batch_size * 10)
+        return dataset
+
+    def get_features_list(self, config, graph, no_of_batches):
+        features_list = []
+        with tf.compat.v1.Session(config=config, graph=graph) as sess:
+            res_dataset = self.input_fn(False)
+            iterator = tf.compat.v1.data.make_one_shot_iterator(res_dataset)
+            next_element = iterator.get_next()
+            for i in range(int(no_of_batches)):
+                batch = sess.run(next_element)
+                features_list.append(batch)
+
+        return features_list
+
+    def unpickle(self, dataset_path):
+        file = bz2.BZ2File(dataset_path, 'rb')
+        features_list = pickle.load(file)
+        file.close()
+        return features_list
+
+    def get_input_array(self):
+
+        try:
+            input_array = self.features_list[self.current_feature][:2]
+        except IndexError:
+            raise utils.OutOfInstances("no more features to process")
+        return input_array
+
+    def reset(self):
+        self.current_feature = 0
+        self.correct = 0
+        return True
+
+    def submit_predictions(self, output_array):
+        predicted_labels = np.argmax(output_array['import/import/head/predictions/probabilities:0'], 1)
+        self.correct = self.correct + np.sum(self.features_list[self.current_feature][2] == predicted_labels)
```

Comment:
Can be also expressed as self.correct += np.sum(self.features_list[self.current_feature][2] == predicted_labels)

Reply:
hmm, what's the np.sum() for? the output of that comparison will be a single bool always so there's nothing to sum together

Reply:
I must have been drunk

Reply:
Actually, self.features_list[self.current_feature][2] and predicted_labels are both numpy arrays so np.sum in this case gives the number of times that the model was right, so it could be a number in hundreds